cpu/stm32f4/periph/adc.c

```diff
@@ -22,164 +22,63 @@
-            break;
-        case ADC_RES_10BIT:
-            adc->CR1 |= ADC_CR1_RES_0;
-            adc_config[dev].max_value = 0x3ff;
-            break;
-        case ADC_RES_12BIT:
-            adc_config[dev].max_value = 0xfff;
-            break;
-        case ADC_RES_14BIT:
-        case ADC_RES_16BIT:
-            adc_poweroff(dev);
-            return -1;
-            break;
-    }
-
-    /* set clock prescaler */
-    ADC->CCR = (3 << 16);       /* ADC clock = 10,5MHz */
-
-    /* enable the ADC module */
-    adc->CR2 |= ADC_CR2_ADON;
-
-    return 0;
+    return adc_config[line].dev;
 }
 
-int adc_sample(adc_t dev, int channel)
+static inline void poweron(adc_t line)
 {
-    ADC_TypeDef *adc = 0;
-
-    switch (dev) {
-#if ADC_0_EN
-        case ADC_0:
-            adc = ADC_0_DEV;
-            switch (channel) {
-                case 0:
-                    adc->SQR3 = ADC_0_CH0 & 0x1f;
-                    break;
-                case 1:
-                    adc->SQR3 = ADC_0_CH1 & 0x1f;
-                    break;
-                default:
-                    return -1;
-            }
-            break;
-#endif
-#if ADC_1_EN
-        case ADC_1:
-            adc = ADC_1_DEV;
-            switch (channel) {
-                case 0:
-                    adc->SQR3 = ADC_1_CH0 & 0x1f;
-                    break;
-                case 1:
-                    adc->SQR3 = ADC_1_CH1 & 0x1f;
-                    break;
-                default:
-                    return -1;
-            }
-            break;
-#endif
-    }
-
-    /* start single conversion */
-    adc->CR2 |= ADC_CR2_SWSTART;
-    /* wait until conversion is complete */
-    while (!(adc->SR & ADC_SR_EOC));
-    /* read and return result */
-    return (int)adc->DR;
+    RCC->APB2ENR |= (1 << adc_config[line].rcc);
 }
 
-void adc_poweron(adc_t dev)
+static inline void poweroff(adc_t line)
 {
-    switch (dev) {
-#if ADC_0_EN
-        case ADC_0:
-            ADC_0_CLKEN();
-            break;
-#endif
-#if ADC_1_EN
-        case ADC_1:
-            ADC_1_CLKEN();
-            break;
-#endif
-    }
+    RCC->APB2ENR &= ~(1 << adc_config[line].rcc);
 }
 
-void adc_poweroff(adc_t dev)
+int adc_init(adc_t line)
 {
-    switch (dev) {
-#if ADC_0_EN
-        case ADC_0:
-            ADC_0_CLKDIS();
-            break;
-#endif
-#if ADC_1_EN
-        case ADC_1:
-            ADC_1_CLKDIS();
-            break;
-#endif
+    /* check if the linenel is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
     }
+    /* configure the pin */
+    gpio_init_adc(adc_config[line].pin);
+    /* power on the device for the duration of configuration */
+    poweron(line);
+    /* set sequence length to 1 conversion and enable the ADC device */
+    dev(line)->SQR1 = 0;
+    dev(line)->CR2 = ADC_CR2_ADON;
+    /* set clock prescaler */
+    ADC->CCR = (3 << 16);       /* ADC clock = 10,5MHz */
+    /* power off the device again */
+    poweroff(line);
```

Comment:
Please note that if you disable the ADC clock you will reset it's configuration to reset state. This means the ADC will be unconfigured next time you call `adc_sample`. Don't know if that was intended?